Cache parse_links() by --find-links html page url

cosmicexplorer opened the pull request:
`pip._internal.index.collector.parse_links()` can take a nontrivial amount of time for large `--find-links` html repos. This cost is paid every single time a `--find-links` page is consulted (once per transitive requirement resolved), and can therefore become quite lengthy for large resolves. See pantsbuild/pex#887 and pantsbuild/pex#888 for further background on how we arrived at this issue while overcoming blockers to https://github.com/pantsbuild/pants upgrading to pex version 2, which uses pip to resolve and fetch wheels.

This change caches the result of invoking `parse_links()` for a `--find-links` html repo, keyed by the `HTMLPage`'s `url`. This means that other code in pip calling `parse_links()` can safely expect to idempotently resolve a set of links from an html page from any other thread, for example, without having to share any state between threads.

To reproduce the pathological behavior and demonstrate that this fix improves performance, check out this branch and execute this script:
```bash
#!/bin/bash

set -euxo pipefail

function generate-large-find-links-page {
  echo '<html><head><meta charset="utf-8"><head><body>'
  set +x
  for i in $(seq 50000); do
    echo '<a href="nonexistent-0.5.0.whl">nonexistent-0.5.0.whl</a>'
  done
  for w in *.whl *.tar.gz; do
    echo "<a href=\"${w}\">${w}</a>"
  done
  set -x
  echo '</body></html>'
}

python setup.py bdist_wheel

mkdir -pv tmp-whls/
pushd tmp-whls/
pip download tensorflow==1.14.0
generate-large-find-links-page > large-find-links-page.html
popd

time PYTHONPATH="$(pwd)/$(echo dist/*.whl)" python -m pip \
     -vvv download --no-index --find-links=./tmp-whls/large-find-links-page.html \
     tensorflow==1.14.0

time pip \
     -vvv download --no-index --find-links=./tmp-whls/large-find-links-page.html \
     tensorflow==1.14.0
```

chrahunt commented:
A decorator would help us keep this logic out of `parse_links`, like

```python
from functools import lru_cache


class CacheablePage(object):
    def __init__(self, page):
        self.page = page

    def __hash__(self):
        return hash((page.content, page.encoding))


def with_cached_html_pages(fn):
    @lru_cache
    def wrapper(cacheable_page):
        return fn(cacheable_page.page)

    def wrapper_wrapper(page):
        return wrapper(CacheablePage(page))

    return wrapper_wrapper

...

@with_cached_html_pages
def parse_links(page):
   ...
```

cosmicexplorer commented:
Thank you -- that sounds great! I will make this change!

uranusjr commented:
Is there a reason behind caching the result against the content, instead of the `find-links` URL/path? I kind of feel the former would be enough.

cosmicexplorer commented:
> Is there a reason behind caching the result against the content, instead of the find-links URL/path? I kind of feel the former would be enough.

Super reasonable! This change was motivated because we saw the repeated message:
```
Request header has "max_age" as 0, cache bypassed
```
which we pinned down to https://github.com/pypa/pip/blob/e633741ea6de70c22704cf24a1ff4eeb9aa5e903/src/pip/_internal/index/collector.py#L130-L143 in pantsbuild/pex#887.

I agree that caching by the url path seems sufficient, and I can definitely see how that might be more "expected" behavior in some respect.

I'm thinking that (with the use case of #5670 in mind) if some build process publishes a change to a `--find-links` repo, we would want to make sure that newly-published wheels aren't cached at this layer of pip. Given that this is non-persistent caching, I can't see any obvious way that this caching keyed on url paths only would cause a regression on #5670, unless perhaps if pip is being invoked as a library by some persistent background process which concurrently performs a publish (which is not a use case I'm aware of, but seems maybe plausible).

But using the url path seems maybe cleaner to implement, and possibly easier to fold into some other caching via `cachecontrol` or elsewhere later on. I think url-based or content-based keying would definitely both fix pantsbuild/pex#887. I'm not sure whether we're safe assuming pip is invoked as a subprocess mostly ephemerally (but I feel like we are?).

uranusjr commented:
> Given that this is non-persistent caching, though, I can't see any obvious way that this caching keyed on url paths only would cause a regression on #5670, unless perhaps if pip is being invoked as a library by some persistent background process which concurrently performs a publish (which is not a use case I'm aware of, but seems maybe plausible).

Yup, I was thinking along a similar line as well, and `lru_cache` being non-persistent is the main reason I feel caching by URI would be correct. pip explicitly does not support being imported as a library, and already relies heavily on global states in other places, so it would be okay to do the same here (IMO). We can always add a cache-busting hook (`lru_cache.cache_clear()`) if by some off-chance we ever want to refactor pip into importable.

cosmicexplorer commented:
> pip explicitly does not support being imported as a library, and already relies heavily on global states in other places

Great! Was being overly cautious! I think the `CacheablePage` keyed on url sounds great for now -- coming right up!

cosmicexplorer commented:
@uranusjr An unexpected consequence of keying by page url is that the testing in `test_collector.py` which invokes `parse_links()` within a `@pytest.mark.parameterized()` now needs to have a unique page url injected to have tests pass (as demonstrated in c399ba248a549b2038afdad8bfe1ace8240d1a28). I believe that keying by page content and encoding avoids this workaround -- but I'm also definitely still ok keying by the url and leaving this small hack in place!

cosmicexplorer commented:
I've detailed in pantsbuild/pip#5 some quick testing I did using the benchmark script in this PR's description on the effects of adding a similar cache for fetching of `HTMLPage`s (*on top of* the `parse_links()` cache), and how it seems that this produces a nonzero speedup even when the `--find-links` html page is available locally (as opposed to being fetched from over a network). I can go into more depth on how I arrived at this conclusion as necessary.

Since the separate `HTMLPage` fetch caching I just added is keyed by url, and because of the above (keying the `parse_links()` cache by url requires adding a `uuid` salt to each url to test), I've also reverted the `CacheablePageContent` cache key for `parse_links()` to use the page's `content` and `encoding` to cache `parse_links()`, not just its url. This was implemented in 8809012.

cosmicexplorer commented:
I've added type annotations for all of the things! I was unable to figure out how to get mypy to accept the:
```python
try:
  from functools import lru_cache
except ImportError:
  def lru_cache(...):
```
, as it would always see both the "try" and the "except" and therefore be unable to satisfy mypy for python 2 and 3. I have therefore created a method `_lru_cache()` which has a clear TODO on it to replace with `@functools.lru_cache()` after dropping python 2 support!

cosmicexplorer commented:
@uranusjr would love to know your thoughts on the change to additionally support caching `HTMLPage` downloads with `@with_cached_link_fetch`, which keys the cache by page url. It appears to have a performance benefit (but I don't have a benchmark offhand except from what was mentioned in https://github.com/pantsbuild/pip/pull/5#issuecomment-586012818).

uranusjr commented:
I would guess the performance benefits (compared to only caching find-links pages) are marginal. Index pages are already per-package, and each package page is only read once in the vast majority of cases (exceptions being build-time dependencies such as setuptools). On the flip side though, the added memory requirements could be significant on constrained systems (cache entry of a well-known package is likely to be in the 100kb-500kb range) that I don’t feel comfortable blindly caching them all without adding a flag to disable it. I would suggest sticking to only caching find-links pages for now.

cosmicexplorer commented:
Thank you for the extremely thoughtful response! I had not considered at all memory-constrained systems, and I agree that the benefits are marginal enough not to warrant these tradeoffs (we do not have any benchmarks showing any further speedup -- this was a mistake on my part, being too eager to find another optimization). I really appreciate you being so specific and helpful in your response! I will revert the change to add an `HTMLPage` fetch cache, and I will revert the cache key for `parse_links()` to key by url!

cosmicexplorer commented:
@uranusjr ping! I have removed the separate caching by page content and have just keyed `parse_links()` by url. I would love to fix any further issues :) no rush, thanks so much for your input so far!

uranusjr commented:
Doesn’t `parse_links()` also handle index pages (e.g. pypi.org)? I am under the impression it does (this could be wrong since I didn’t actually read the implementation), and I don’t really want to cache them.

cosmicexplorer commented:
Thank you! I will check that and avoid caching index pages if so!

cosmicexplorer commented:
@uranusjr in beac52ad08ef8008b010eebee7aad2f9dd21d99f I have marked `Link` objects coming from index urls with a new instance field `is_index_url` for `Link`, which is then transferred to the `HTMLPage` object created from the `Link` in `_get_html_page()`. I have then modified the `@with_cached_html_pages` wrapper to avoid using the cache if `is_index_url=True`.

Since I had to modify a couple other files to add `is_index_url`, it might make sense to add a more "end-to-end" test which will actually exercise the code path where we construct `Link(url, is_index_url=True)`. I'll see if there are any good examples of that in the codebase already.

cosmicexplorer commented:
@uranusjr ping! I have added appropriate testing and a couple docstrings.

cosmicexplorer commented:
Ping again!

uranusjr commented:
Hi! Sorry I’ve been a bit tied up recently, I will find some time to review this entirely this weekend.

cosmicexplorer commented:
Thanks!! Not a problem!! No rush on my end ^_^ Thanks so much for your efforts to make this change work!

uranusjr commented on news/7729.bugfix:
```diff
@@ -0,0 +1,3 @@
+Cache the result of parse_links() to avoid re-tokenizing a find-links page multiple times over a pip run.
+
+This change significantly improves resolve performance when --find-links points to a very large html page.
```
I’m wondering from reading this text whether this should be a `.feature` instead 🤔

cosmicexplorer replied:
True, it's also not really a bug at all to avoid caching here -- the caching is just something that pex v1 happened to do (possibly without realizing it), which then was really exposed as a feature (as you've mentioned) when pex v2 started using pip.

cosmicexplorer replied:
I have renamed it to `7729.feature` pending further discussion!

uranusjr commented on src/pip/_internal/index/collector.py:
```diff
@@ -42,6 +43,35 @@
+            cache[cache_key] = value
+            return value
+        return wrapped
+    return wrapper
```
I wonder whether we should simply make the cache a no-op on Python 2. Something like:

```python
F = TypeVar('F', Callable)

def stub_lru_cache(maxsize=None):
    # type: (Optional[int]) -> Callable[[F], F]
    def _wrapper(f):
        # type: (F) -> F
        return f
    return _wrapper
try:
    from functools import lru_cache as real_lru_cache
except ImportError:
    _lru_cache = stub_lru_cache
else:
    _lru_cache = real_lru_cache
```

There are probably a couple of typing hoops we need to jump through, but the point is not avoid vendoring a duplicated `lru_cache` altogether.

uranusjr replied:
A couple of references for convinience, if we are going down this route:

* python/mypy#1297 How to deal with conditional imports
* python/mypy#5107 How to type-check `lru_cache`

cosmicexplorer replied:
I have slightly modified your example to make mypy and flake8 accept it: https://github.com/pypa/pip/blob/83153538e61e056b9b99f5334684e7001081a90a/src/pip/_internal/index/collector.py#L49-L57

uranusjr replied:
I pushed a patch against your repo to replace the `type: ignore` comments with `getattr()` and a callback protocol that IMO is easier to follow. You can merge that PR to include it in this PR.

cosmicexplorer replied:
Thank you for diving into this and figuring out the Protocol usage!! I have merged the PR!

xavfernandez commented on src/pip/_internal/index/collector.py:
```diff
@@ -38,10 +39,29 @@
+        return f
+    return _wrapper
+
+
```
A comment mentioning Python 2 ("Fallback to noop_lru_cache in Python 2") would increase the chances we'll clean this when dropping python 2 support

cosmicexplorer replied:
Yes, absolutely. This was lost when I refactored it. Will add!

xavfernandez commented on tests/unit/test_collector.py:
```diff
@@ -355,14 +357,63 @@ def test_parse_links__yanked_reason(anchor_html, expected):
+@skip_if_python2
+def test_parse_links_caches_same_page_by_url():
+    html = (
+        # Mark this as a unicode string for Python 2 since anchor_html
```
This comment (and the `u''` prefix) aren't relevant in a `@skip_if_python2` test ^^

cosmicexplorer replied:
Thank you!!! Will fix!

cosmicexplorer commented:
> If that's the case, a simple cache_link_parsing=False argument might make more sense ?

Absolutely!! I'll make this change!

xavfernandez commented:
@pradyunsg this should be ready for version 20.1

pradyunsg commented:
@xavfernandez feel free to click the merge button! :)

xavfernandez commented:
Thanks @cosmicexplorer :)
